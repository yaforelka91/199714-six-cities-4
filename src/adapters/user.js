const adaptUser = (user) => {
  return {
    id: user.id,
    email: user.email,
    name: user.name,
    picture: user.avatar_url,
    isSuper: user.is_pro,
  };
};

export default adaptUser;
