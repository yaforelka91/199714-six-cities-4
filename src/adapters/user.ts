import User from '../interfaces/user';

const adaptUser = (user): User => ({
  id: user.id,
  email: user.email,
  name: user.name,
  picture: user.avatar_url,
  isSuper: user.is_pro,
});

export default adaptUser;
