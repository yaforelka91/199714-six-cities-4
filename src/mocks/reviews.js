export default {
  '0': {
    id: 0,
    user: {
      name: `Max`,
      picture: `http://placekitten.com/54/54`,
    },
    rating: 4.2,
    feedback: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
    visitTime: `2019-04-24`,
  },
  '1': {
    id: 1,
    user: {
      name: `Maria`,
      picture: `http://placekitten.com/54/54`,
    },
    rating: 5,
    feedback: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
    visitTime: `2018-05-25`,
  },
  '2': {
    id: 2,
    user: {
      name: `Mike`,
      picture: `http://placekitten.com/54/54`,
    },
    rating: 4,
    feedback: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
    visitTime: `2019-01-01`,
  },
  '3': {
    id: 3,
    user: {
      name: `Alice`,
      picture: `http://placekitten.com/54/54`,
    },
    rating: 4.7,
    feedback: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
    visitTime: `2016-12-30`,
  },
  '4': {
    id: 4,
    user: {
      name: `John`,
      picture: `http://placekitten.com/54/54`,
    },
    rating: 4.5,
    feedback: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam. The building is green and from 18th century.`,
    visitTime: `2020-03-08`,
  },
};
