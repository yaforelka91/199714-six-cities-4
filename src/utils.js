import {createSelector} from 'reselect';
import {SortType, CardView, MONTHS_LIST} from './const.js';

export const extend = (object1, object2) => {
  return Object.assign({}, object1, object2);
};

export const sortOffers = (sortType, offers) => {
  switch (sortType) {
    case SortType.POPULAR:
      return offers.slice();
    case SortType.TO_HIGH:
      return offers.slice().sort((prevOffer, nextOffer) => prevOffer.price - nextOffer.price);
    case SortType.TO_LOW:
      return offers.slice().sort((prevOffer, nextOffer) => nextOffer.price - prevOffer.price);
    case SortType.TOP_RATED:
      return offers.slice().sort((prevOffer, nextOffer) => nextOffer.rating - prevOffer.rating);
    default:
      return offers;
  }
};

export const capitalize = (word) => {
  return word.charAt(0).toUpperCase() + word.slice(1);
};

export const getFormattedDate = (dateStr) => {
  const date = new Date(dateStr);
  return `${MONTHS_LIST[date.getMonth()]} ${date.getFullYear()}`;
};

export const getSortedOffers = createSelector(
    (state) => {
      return state.sortType;
    },
    (state) => {
      return state.offers;
    },
    (activeSorting, offers) => {
      return sortOffers(activeSorting, offers);
    }
);

export const getRatingInPercent = (ratingValue) => {
  return Math.round(ratingValue * 100 / 5);
};

export const getArticleClassName = (viewMode) => {
  switch (viewMode) {
    case CardView.CITIES:
      return {
        classNameForArticle: `cities__place-card`,
        classNameForImage: `cities__image-wrapper`,
        classNameForInfo: ``,
      };

    case CardView.NEAR:
      return {
        classNameForArticle: `near-places__card`,
        classNameForImage: `near-places__image-wrapper`,
        classNameForInfo: ``,
      };

    case CardView.FAVORITES:
      return {
        classNameForArticle: `favorites__card`,
        classNameForImage: `favorites__image-wrapper`,
        classNameForInfo: `favorites__card-info`,
      };
  }
  return {
    classNameForArticle: ``,
    classNameForImage: ``,
    classNameForInfo: ``,
  };
};
