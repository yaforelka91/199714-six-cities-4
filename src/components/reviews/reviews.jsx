import React from 'react';
import ReviewCard from '../review-card/review-card.jsx';
import ReviewForm from '../review-form/review-form.jsx';
import {reviewsTypes} from '../../types/types.js';
import {AuthorizationStatus} from '../../reducer/user/user.js';
import {connect} from 'react-redux';
import {Operation} from '../../reducer/reviews/reviews.js';
import {getAuthorizationStatus} from '../../reducer/user/selectors.js';
import withReview from '../../hocs/with-review/with-review.js';
import {getReviewsCount, getReducedReviews} from '../../reducer/reviews/selectors.js';

const ReviewFormWrapped = withReview(ReviewForm);
const Reviews = ({authorizationStatus, reviewsCount, reviews, offerId, className, onReviewFormSubmit}) => {
  return (
    <section className={`review${className ? ` ${className}` : ``}`}>
      <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{reviewsCount}</span></h2>
      {reviewsCount > 0 ?
        <ul className="reviews__list">
          {
            reviews
            .map((review) => {
              return (
                <ReviewCard key={review.id} review={review} offerId={offerId} />
              );
            })
          }
        </ul>
        : null
      }
      {
        authorizationStatus === AuthorizationStatus.AUTH &&
      <ReviewFormWrapped
        onReviewFormSubmit={onReviewFormSubmit}
        offerId={offerId}
      />
      }
    </section>
  );
};

Reviews.defaultProps = {
  className: ``,
};

Reviews.propTypes = reviewsTypes;

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  reviewsCount: getReviewsCount(state),
  reviews: getReducedReviews(state),
});

const mapDispatchToProps = (dispatch) => ({
  onReviewFormSubmit(commentData, hotelId) {
    return dispatch(Operation.sendComment(commentData, hotelId));
  }
});

export {Reviews};
export default connect(mapStateToProps, mapDispatchToProps)(Reviews);
