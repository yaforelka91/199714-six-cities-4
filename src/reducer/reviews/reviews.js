import {extend} from '../../utils.js';
import adaptReview from '../../adapters/review.js';

const initialState = {
  reviews: [],
};

const ActionType = {
  SEND_REVIEW: `SEND_REVIEW`,
  LOAD_REVIEWS: `LOAD_REVIEWS`,
};

const ActionCreator = {
  sendReview: (reviewData) => ({
    type: ActionType.SEND_REVIEW,
    payload: reviewData,
  }),
  loadReviews: (reviews) => ({
    type: ActionType.LOAD_REVIEWS,
    payload: reviews,
  }),
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.SEND_REVIEW:
      return extend(state, {
        reviews: action.payload,
      });
    case ActionType.LOAD_REVIEWS:
      return extend(state, {
        reviews: action.payload,
      });
  }

  return state;
};

const Operation = {
  loadReviews: (hotelId) => (dispatch, getState, api) => {
    return api.get(`/comments/${hotelId}`)
      .then((response) => {
        const adaptedReviews = response.data.map((review) => adaptReview(review));
        dispatch(ActionCreator.loadReviews(adaptedReviews));
      });
  },
  sendComment: (commentData, hotelId) => (dispatch, getState, api) => {
    return api.post(`/comments/${hotelId}`, {
      comment: commentData.comment,
      rating: commentData.rating,
    })
      .then((response) => {
        const adaptedReviews = response.data.map((review) => adaptReview(review));

        dispatch(ActionCreator.sendReview(adaptedReviews));
      })
    .catch((err) => {
      throw err;
    });
  },
};

export {
  ActionCreator,
  ActionType,
  Operation,
  reducer,
};
