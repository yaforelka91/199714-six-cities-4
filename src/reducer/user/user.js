import {extend} from '../../utils';
import adaptUser from '../../adapters/user';
import adaptError from '../../adapters/error';
import {Operation as DataOperation} from '../data/data';
import {Error} from '../../api';

const AuthorizationStatus = {
  AUTH: `AUTH`,
  NO_AUTH: `NO_AUTH`,
};

const initialState = {
  authorizationStatus: AuthorizationStatus.NO_AUTH,
  userData: {},
  isAuthorizationInProgress: true,
  errorType: ``,
};

const ActionType = {
  REQUIRED_AUTHORIZATION: `REQUIRED_AUTHORIZATION`,
  SET_USER_DATA: `SET_USER_DATA`,
  CHANGE_PROGRESS_STATUS: `CHANGE_PROGRESS_STATUS`,
  CATCH_SERVER_ERROR: `CATCH_SERVER_ERROR`,
};

const ActionCreator = {
  requireAuthorization: (status) => ({
    type: ActionType.REQUIRED_AUTHORIZATION,
    payload: status,
  }),
  setUserData: (authData) => ({
    type: ActionType.SET_USER_DATA,
    payload: authData,
  }),
  changeProgressStatus: (isLoaded) => ({
    type: ActionType.CHANGE_PROGRESS_STATUS,
    payload: isLoaded,
  }),
  catchServerError: (errorMessage) => ({
    type: ActionType.CATCH_SERVER_ERROR,
    payload: errorMessage,
  }),
};

const reducer = (state = initialState, action) => {
  switch (action.type) {
    case ActionType.REQUIRED_AUTHORIZATION:
      return extend(state, {
        authorizationStatus: action.payload,
      });

    case ActionType.SET_USER_DATA:
      return extend(state, {
        userData: action.payload,
      });

    case ActionType.CHANGE_PROGRESS_STATUS:
      return extend(state, {
        isAuthorizationInProgress: action.payload,
      });

    case ActionType.CATCH_SERVER_ERROR:
      return extend(state, {
        errorType: action.payload,
      });
  }

  return state;
};

const Operation = {
  checkAuth: () => (dispatch, getState, api) => {
    return api.get(`/login`)
      .then((response) => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.setUserData(adaptUser(response.data)));
        dispatch(ActionCreator.changeProgressStatus(false));
      })
      .catch((err) => {
        dispatch(ActionCreator.changeProgressStatus(false));
        throw err;
      });
  },
  login: (authData) => (dispatch, getState, api) => {
    return api.post(`/login`, {
      email: authData.login,
      password: authData.password,
    })
      .then((response) => {
        dispatch(ActionCreator.requireAuthorization(AuthorizationStatus.AUTH));
        dispatch(ActionCreator.setUserData(adaptUser(response.data)));
        dispatch(ActionCreator.catchServerError(``));
        dispatch(DataOperation.loadOffers());
        dispatch(ActionCreator.changeProgressStatus(false));
      })
      .catch((err) => {
        if (err.response.status === Error.BAD_REQUEST) {
          dispatch(ActionCreator.catchServerError(adaptError(err.response.data.error)));
          dispatch(ActionCreator.changeProgressStatus(false));
          return;
        }
        dispatch(ActionCreator.changeProgressStatus(false));
        throw err;
      });
  },
};

export {
  ActionCreator,
  ActionType,
  AuthorizationStatus,
  Operation,
  reducer,
};
