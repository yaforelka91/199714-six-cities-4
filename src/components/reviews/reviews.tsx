import React from 'react';
import ReviewCard from '../review-card/review-card';
import ReviewForm from '../review-form/review-form';
import {AuthorizationStatus} from '../../reducer/user/user';
import {connect} from 'react-redux';
import {Operation} from '../../reducer/reviews/reviews';
import {getAuthorizationStatus} from '../../reducer/user/selectors';
import withReview from '../../hocs/with-review/with-review';
import {getReviewsCount, getSortedReviews} from '../../reducer/reviews/selectors';
import Review from '../../interfaces/review';

const MAX_COUNT_REVIEWS = 10;
const ReviewFormWrapped = withReview(ReviewForm);

type Props = {
  reviews: Review[];
  reviewsCount: number;
  authorizationStatus: string;
  offerId: number;
  className?: string;
  onReviewFormSubmit: (commentData: {rating: number; comment: string}, hotelId: number) => Promise<void>;
}

const Reviews: React.FC<Props> = (props: Props) => {
  const {
    authorizationStatus,
    reviewsCount,
    reviews,
    offerId,
    className = ``,
    onReviewFormSubmit
  } = props;

  return (
    <section className={`review${className ? ` ${className}` : ``}`}>
      <h2 className="reviews__title">Reviews &middot; <span className="reviews__amount">{reviewsCount}</span></h2>
      {reviewsCount > 0 &&
        <ul className="reviews__list">
          {
            reviews.map((review) => <ReviewCard key={review.id} review={review} />)
          }
        </ul>
      }
      {
        authorizationStatus === AuthorizationStatus.AUTH &&
      <ReviewFormWrapped
        onReviewFormSubmit={onReviewFormSubmit}
        offerId={offerId}
      />
      }
    </section>
  );
};

const mapStateToProps = (state) => ({
  authorizationStatus: getAuthorizationStatus(state),
  reviewsCount: getReviewsCount(state),
  reviews: getSortedReviews(state).slice(0, MAX_COUNT_REVIEWS),
});

const mapDispatchToProps = (dispatch) => ({
  onReviewFormSubmit(commentData, hotelId) {
    return dispatch(Operation.sendComment(commentData, hotelId));
  }
});

export {Reviews};
export default connect(mapStateToProps, mapDispatchToProps)(Reviews);
