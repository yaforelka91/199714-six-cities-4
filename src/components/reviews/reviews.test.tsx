import * as React from 'react';
import * as renderer from 'react-test-renderer';
import {Reviews} from './reviews';
import {AuthorizationStatus} from '../../reducer/user/user.js';
import {noOperation} from '../../utils';
import Review from '../../interfaces/review';

const reviews: Review[] = [
  {
    id: 2,
    user: {
      name: `Maria`,
      picture: `pic`,
    },
    rating: 5,
    feedback: `Text`,
    visitTime: `2019-05-25`,
  },
  {
    id: 1,
    user: {
      name: `Max`,
      picture: `pic`,
    },
    rating: 4.2,
    feedback: `Text`,
    visitTime: `2018-04-24`,
  },
];

describe(`ReviewsSnapshot`, () => {
  it(`should render Review with review form`, () => {
    const tree = renderer.create(
        <Reviews
          reviews={reviews}
          reviewsCount={2}
          authorizationStatus={AuthorizationStatus.AUTH}
          offerId={1}
          onReviewFormSubmit={noOperation}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });

  it(`should render Review without review form`, () => {
    const tree = renderer.create(
        <Reviews
          reviews={reviews}
          reviewsCount={2}
          authorizationStatus={AuthorizationStatus.NO_AUTH}
          offerId={1}
          onReviewFormSubmit={noOperation}
        />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
