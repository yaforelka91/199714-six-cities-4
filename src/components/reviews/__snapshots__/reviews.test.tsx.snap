// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ReviewsSnapshot should render Review with review form 1`] = `
<section
  className="review"
>
  <h2
    className="reviews__title"
  >
    Reviews · 
    <span
      className="reviews__amount"
    >
      2
    </span>
  </h2>
  <ul
    className="reviews__list"
  >
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="pic"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Maria
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "100%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          Text
        </p>
        <time
          className="reviews__time"
          dateTime="2019-05-25"
        >
          May 2019
        </time>
      </div>
    </li>
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="pic"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Max
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "80%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          Text
        </p>
        <time
          className="reviews__time"
          dateTime="2018-04-24"
        >
          April 2018
        </time>
      </div>
    </li>
  </ul>
  <form
    action="#"
    className="reviews__form form"
    method="post"
    onSubmit={[Function]}
  >
    <label
      className="reviews__label form__label"
      htmlFor="review"
    >
      Your review
    </label>
    <div
      className="reviews__rating-form form__rating"
    >
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="5-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value="5"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="5-stars"
        title="perfect"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="4-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value="4"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="4-stars"
        title="good"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="3-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value="3"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="3-stars"
        title="not bad"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="2-stars"
        name="rating"
        onChange={[Function]}
        type="radio"
        value="2"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="2-stars"
        title="badly"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
      <input
        checked={false}
        className="form__rating-input visually-hidden"
        id="1-star"
        name="rating"
        onChange={[Function]}
        type="radio"
        value="1"
      />
      <label
        className="reviews__rating-label form__rating-label"
        htmlFor="1-star"
        title="terribly"
      >
        <svg
          className="form__star-image"
          height="33"
          width="37"
        >
          <use
            xlinkHref="#icon-star"
          />
        </svg>
      </label>
    </div>
    <textarea
      className="reviews__textarea form__textarea"
      id="review"
      maxLength={300}
      name="review"
      onChange={[Function]}
      placeholder="Tell how was your stay, what you like and what can be improved"
      value=""
    />
    <div
      className="reviews__button-wrapper"
    >
      <p
        className="reviews__help"
      >
        To submit review please make sure to set 
        <span
          className="reviews__star"
        >
          rating
        </span>
         and describe your stay with at least 
        <b
          className="reviews__text-amount"
        >
          50 characters
        </b>
        .
      </p>
      <button
        className="reviews__submit form__submit button"
        disabled={true}
        type="submit"
      >
        Submit
      </button>
    </div>
    
  </form>
</section>
`;

exports[`ReviewsSnapshot should render Review without review form 1`] = `
<section
  className="review"
>
  <h2
    className="reviews__title"
  >
    Reviews · 
    <span
      className="reviews__amount"
    >
      2
    </span>
  </h2>
  <ul
    className="reviews__list"
  >
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="pic"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Maria
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "100%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          Text
        </p>
        <time
          className="reviews__time"
          dateTime="2019-05-25"
        >
          May 2019
        </time>
      </div>
    </li>
    <li
      className="reviews__item"
    >
      <div
        className="reviews__user user"
      >
        <div
          className="reviews__avatar-wrapper user__avatar-wrapper"
        >
          <img
            alt="Reviews avatar"
            className="reviews__avatar user__avatar"
            height="54"
            src="pic"
            width="54"
          />
        </div>
        <span
          className="reviews__user-name"
        >
          Max
        </span>
      </div>
      <div
        className="reviews__info"
      >
        <div
          className="reviews__rating rating"
        >
          <div
            className="reviews__stars rating__stars"
          >
            <span
              style={
                Object {
                  "width": "80%",
                }
              }
            />
            <span
              className="visually-hidden"
            >
              Rating
            </span>
          </div>
        </div>
        <p
          className="reviews__text"
        >
          Text
        </p>
        <time
          className="reviews__time"
          dateTime="2018-04-24"
        >
          April 2018
        </time>
      </div>
    </li>
  </ul>
</section>
`;
